### PR TITLE
#454 Add service file functions to update `participant_activities` completion status

### LINKED ISSUE #454: programsService should support updating of ParticipantActivities completion status
## Expected behaviour

There should be service functions that marks a target activity as completed once a user click the complete button on an Activity Dialog. And also having the corresponding test to make sure the function works correctly.

## Actual behaviour

It currently doesn't exist.

#### This ticket associated with the other backend router ticket #455 is to provide functions for the frontend ticket (#...) to produce a feature that user can mark any activity as completed.
- A `getParticipantActivityCompletion` function returns the `completed` (a boolean value, true or false)
- A `getParticipantActivityId` checks if there is an activity that matches the given `program_id`, `activity_id `, `principal_id` from table `participant_activities`.  If yes, returns `participantActivityId`. If not, respectively insert rows of `program_id`, `activity_id `, `principal_id` with corresponding value and returns `participantActivityId`.
- A `setParticipantActivityCompletion` function toggles(updates) the `completed` field based on the given `participant_activities ID`, `program_id`, `activity_id `, `principal_id`, `completed`.

****
    - If the row does not exist in the table, return false.
    - If the row does exist in the table, return the value of the 'completed' attribute
****
## Checklist

- [x] This issue is linked to the appropriate project.
- [x] This issue has appropriate labels.

### PR DESCRIPTION
**Pull request by phoebehala**
## Proposed changes
This resolves  #454 by:
- A `getParticipantActivityId` function returns the participant activity ID if it exists, otherwise it returns null.
- A `getParticipantActivityCompletion` function returns the completion status (either true or false) of the specific activity.
- A `setParticipantActivityCompletion` function is to set/update the completion status of the specific activity.
-  The JsDoc function headers for each of functions above.
-  The JEST test functions for each of functions above.

## Checklist

- [x] Are the issues being addressed linked to this PR?
- [x] Do all commit messages start with the issue number?
- [x] Are all code changes sufficiently tested?
- [ ] Are there screenshots for UI changes?

### CODE REVIEW

**Comment by ChainShoe-Z** [comment]
All tests are passed, just remove the comment and run delint again!

**Review by daveytech — CHANGES_REQUESTED** [review]
Great work on this ticket. One comment left on the file that should probably be cleaned up. And then a question about the choice on the knex query. Other than that should be good for a merge!

File: api/src/services/programsService.ts [hunk 1]
```diff
@@ -289,3 +289,88 @@ export const listProgramsWithActivities = async () => {
...
+) => {
+  const [participantActivityRow] = await db('participant_activities')
+    .insert({
+      principal_id: principalId,
+      program_id: programId,
+      activity_id: activityId,
+      completed: completed,
+    })
+    .onConflict(['program_id', 'principal_id', 'activity_id'])
+    .merge({ completed: completed });
```
[the remark this answers is not in the document]
Comment (phoebehala) [reply api/src/services/programsService.ts:371]: `.onConflict()` if conflict occurs after inserting a new row which means it already has a row matching 'program_id', 'principal_id', 'activity_id'),
then`.merge({})` update with different data to the insert( update 'completed' value to what we receive from a user)

Refer to https://knexjs.org/guide/query-builder.html#insert

[the remark this answers is not in the document]
Comment (daveytech) [reply api/src/services/programsService.ts:371]: Thank you for the context!

**Review by daveytech — APPROVED** [review]
Thanks for cleaning that up!

File: api/src/models.d.ts [hunk 2]
```diff
@@ -54,3 +54,11 @@ export interface ActivityType {
   created_at?: string;
   updated_at?: string;
 }
+
+export interface participantActivities {
```
Comment (wendyyng) [inline api/src/models.d.ts:58]: I think participantActivities should be capitalized like other types

File: api/src/services/programsService.ts [hunk 3]
```diff
@@ -289,3 +289,88 @@ export const listProgramsWithActivities = async () => {
...
+/**
+ * Set/Update the completion status of the specific activity. If the
+ * `participant_activities` record doesn't exist, create it.
+ *
+ * @param {number} principalId - the unique id for the user
+ * @param {number} programId - the id for the unique program
+ * @param {number} activityId - the id for the unique activity
+ * @param {boolean} completed - the boolean value that a user wants to set
+ *    into the completed field
+ * @returns {Object} - if successful, the ID of the record in the table
```
Comment (rrabinovitch) [inline api/src/services/programsService.ts:354]: same question here as what I commented on L324

File: api/src/services/__tests__/programsService.ts [hunk 4]
```diff
@@ -9,13 +9,17 @@ import {
...
+  getParticipantActivityCompletion,
+  setParticipantActivityCompletion,
 } from '../programsService';
 import {
   Program,
   ProgramActivity,
   CurriculumActivity,
   ActivityType,
   ProgramWithActivities,
+  participantActivities,
```
Comment (rrabinovitch) [inline api/src/services/__tests__/programsService.ts:22]: when Wendy's comment in the models file is addressed, the capitalization should be addressed here as well

**Review by rrabinovitch — CHANGES_REQUESTED** [review]
Great work on this ticket. I know there were a lot of back and forths to get to this point.
I left a few comments that are hopefully quick to address. Other than that, the tests all run successfully on my end.

File: api/src/services/programsService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (seidior) [reply api/src/services/programsService.ts:372]: @daveytech This was my suggestion. It's the officially-sanctioned way in Knex to do an upsert.

**Review by rrabinovitch — APPROVED** [review]

**Review by rrabinovitch — APPROVED** [review]
thanks for addressing my previous comments. it looks like this branch needs to be updated to align with the latest version of main and then hopefully it's ready to merge the PR.

**Pull request merged**